[CMake] respect LLVMConfig.cmake's LLVM_DEFINITIONS in standalone builds

jeremyd2019 opened the pull request:
In #138329, _GNU_SOURCE was added for Cygwin, but when building Clang standalone against an installed LLVM this definition was not picked up, resulting in undefined strnlen.  Follow the documentation in https://llvm.org/docs/CMake.html#embedding-llvm-in-your-project and add the LLVM_DEFINITIONS in standalone projects' cmakes.

jeremyd2019 committed:
[CMake] respect LLVMConfig.cmake's LLVM_DEFINITIONS

In #138329, _GNU_SOURCE was added for Cygwin, but when building Clang
standalone against an installed LLVM this definition was not picked up,
resulting in undefined strnlen.  Follow the documentation in
https://llvm.org/docs/CMake.html#developing-llvm-passes-out-of-source
and add the LLVM_DEFINITIONS in standalone projects' cmakes.

jeremyd2019 commented:
If this is right, it should probably be done to other standalone-capable projects' CMakeLists.txt also (LLD in particular, for my interests)

mstorsjo commented:
> If this is right, it should probably be done to other standalone-capable projects' CMakeLists.txt also (LLD in particular, for my interests). Actually, it seems there's nothing in LLD that requires _GNU_SOURCE on Cygwin...

Yep, indeed.

I guess the main question is who others regularly build in standalone mode and is familiar with how this is supposed to work, cmake-wise. CC @mgorny @tstellar who seem to have touched things for standalone build mode, and CC @petrhosek for overall cmake review.

petrhosek commented:
I'd slightly prefer doing this as two separate changes, that is the first change to replace the use of `list(REMOVE ...)` with `CMakePushCheckState` and the second change to use `LLVMConfig.cmake`'s `LLVM_DEFINITIONS` since these are technically orthogonal.

I agree that the second change, that is using `LLVMConfig.cmake`'s `LLVM_DEFINITIONS`, should be applied to other projects that support standalone builds like LLD.

mstorsjo commented:
> I'd slightly prefer doing this as two separate changes, that is the first change to replace the use of `list(REMOVE ...)` with `CMakePushCheckState` and the second change to use `LLVMConfig.cmake`'s `LLVM_DEFINITIONS` since these are technically orthogonal.

Sounds reasonable!

> I agree that the second change, that is using `LLVMConfig.cmake`'s `LLVM_DEFINITIONS`, should be applied to other projects that support standalone builds like LLD.

Ok, so this is an oversight, and it just happens that none of the projects that support standalone builds have been built in configurations where the defines in `LLVM_DEFINITIONS` have been essential so far?

jeremyd2019 commented:
> I agree that the second change, that is using `LLVMConfig.cmake`'s `LLVM_DEFINITIONS`, should be applied to other projects that support standalone builds like LLD.

Is there a list somewhere or do I just rely on my grep-foo?
* bolt
* clang
* flang
* lld
* lldb
* mlir

grep also turned up these, but they don't seem like things that link to LLVM, so probably shouldn't have these defines.
* compiler-rt
* libc
* libclc
* llvm-libgcc

jeremyd2019 commented:
Also, should all of those projects be updated in one pull request, or do I need to open a half-dozen more?

mstorsjo commented:
> > I agree that the second change, that is using `LLVMConfig.cmake`'s `LLVM_DEFINITIONS`, should be applied to other projects that support standalone builds like LLD.
> 
> Is there a list somewhere or do I just rely on my grep-foo?
> 
> * bolt
> * clang
> * flang
> * lld
> * lldb
> * mlir
> * polly
> 
> grep also turned up these, but they don't seem like things that link to LLVM, so probably shouldn't have these defines.
> 
> * compiler-rt
> * libc
> * libclc
> * llvm-libgcc

Yes, this seems correct.

> Also, should all of those projects be updated in one pull request, or do I need to open a half-dozen more?

I think it's fine to do that kind of project-wide consistency fixes all in one pull request, especially if the change in each of the projects isn't very large and involved.

mgorny commented:
Yeah, I think it's a good idea. Thanks for doing that!

jeremyd2019 commented:
I rebased this on top of #138783 and adjusted the title and description.  Now it should be in a good state to push cmake changes for other projects.

jeremyd2019 committed:
fixup! [CMake] respect LLVMConfig.cmake's LLVM_DEFINITIONS

bolt

jeremyd2019 committed:
fixup! [CMake] respect LLVMConfig.cmake's LLVM_DEFINITIONS

flang

jeremyd2019 committed:
fixup! [CMake] respect LLVMConfig.cmake's LLVM_DEFINITIONS

lld

jeremyd2019 committed:
fixup! [CMake] respect LLVMConfig.cmake's LLVM_DEFINITIONS

lldb

jeremyd2019 committed:
fixup! [CMake] respect LLVMConfig.cmake's LLVM_DEFINITIONS

mlir

jeremyd2019 committed:
fixup! [CMake] respect LLVMConfig.cmake's LLVM_DEFINITIONS

polly

jeremyd2019 commented:
> > I rebased this on top of #138783 and adjusted the title and description. Now it should be in a good state to push cmake changes for other projects.
> 
> The changes look good, but it looks like the changes from #138783 still show up when viewing the changes; can you check that you've rebased past the merged #138783?

I had not - I have now though.

> (Also, I take it that no other subprojects than clang need the `cmake_push_check_state` change?)

No, the other projects were not messing with `_GNU_SOURCE` like clang was.

mstorsjo reviewed (approved):
LGTM

IIRC @petrhosek had commented on this before, and was generally in favour of it, but I'd still leave it open for a couple days if he wants to comment further on it.
